### ISSUE TITLE
'secret_token' not set when creating new user during installation

### ISSUE THREAD

**Issue by dorianmuthig** [alexweissman/UserFrosting#450]
The title says it all, I believe. I was trying to use the latest stable release (v0.3.1).  I hope this will help you fix it.

```php
Slim Applcation Error

Type: Illuminate\Database\QueryException
Code: HY000
Message: SQLSTATE[HY000]: General error: 1364 Field 'secret_token' doesn't have a default value (SQL: insert into `user` (`user_name`, `display_name`, `email`, `password`, `flag_verified`, `locale`, `primary_group_id`, `title`, `updated_at`, `created_at`) values (dorianmuthig, Dorian Muthig, contact@dorianmuthig.com, ?$!password_hash!, 1, en_US, 1, New User, 2015-11-04 14:43:07, 2015-11-04 14:43:07))
File: $(SolutionDir)\userfrosting\vendor\illuminate\database\Connection.php
Line: 624
Trace:
#0 $(SolutionDir)\userfrosting\vendor\illuminate\database\Connection.php(580): Illuminate\Database\Connection->runQueryCallback('insert into `us...', Array, Object(Closure))
#1 $(SolutionDir)\userfrosting\vendor\illuminate\database\Connection.php(363): Illuminate\Database\Connection->run('insert into `us...', Array, Object(Closure))
#2 $(SolutionDir)\userfrosting\vendor\illuminate\database\Connection.php(320): Illuminate\Database\Connection->statement('insert into `us...', Array)
#3 $(SolutionDir)\userfrosting\vendor\illuminate\database\Query\Processors\Processor.php(30): Illuminate\Database\Connection->insert('insert into `us...', Array)
#4 $(SolutionDir)\userfrosting\vendor\illuminate\database\Query\Builder.php(1725): Illuminate\Database\Query\Processors\Processor->processInsertGetId(Object(Illuminate\Database\Query\Builder), 'insert into `us...', Array, 'id')
#5 [internal function]: Illuminate\Database\Query\Builder->insertGetId(Array, 'id')
#6 $(SolutionDir)\userfrosting\vendor\illuminate\database\Eloquent\Builder.php(933): call_user_func_array(Array, Array)
#7 $(SolutionDir)\userfrosting\vendor\illuminate\database\Eloquent\Model.php(1609): Illuminate\Database\Eloquent\Builder->__call('insertGetId', Array)
#8 $(SolutionDir)\userfrosting\vendor\illuminate\database\Eloquent\Model.php(1609): Illuminate\Database\Eloquent\Builder->insertGetId(Array, 'id')
#9 $(SolutionDir)\userfrosting\vendor\illuminate\database\Eloquent\Model.php(1579): Illuminate\Database\Eloquent\Model->insertAndSetId(Object(Illuminate\Database\Eloquent\Builder), Array)
#10 $(SolutionDir)\userfrosting\vendor\illuminate\database\Eloquent\Model.php(1485): Illuminate\Database\Eloquent\Model->performInsert(Object(Illuminate\Database\Eloquent\Builder), Array)
#11 $(SolutionDir)\userfrosting\models\database\User.php(417): Illuminate\Database\Eloquent\Model->save(Array)
#12 $(SolutionDir)\userfrosting\controllers\InstallController.php(236): UserFrosting\User->save()
#13 $(ProjectDir)\index.php(349): UserFrosting\InstallController->setupMasterAccount()
#14 [internal function]: {closure}('master')
#15 $(SolutionDir)\userfrosting\vendor\slim\slim\Slim\Route.php(468): call_user_func_array(Object(Closure), Array)
#16 $(SolutionDir)\userfrosting\vendor\slim\slim\Slim\Slim.php(1357): Slim\Route->dispatch()
#17 $(SolutionDir)\userfrosting\vendor\slim\slim\Slim\Middleware\Flash.php(85): Slim\Slim->call()
#18 $(SolutionDir)\userfrosting\vendor\slim\slim\Slim\Middleware\MethodOverride.php(92): Slim\Middleware\Flash->call()
#19 $(SolutionDir)\userfrosting\middleware\CsrfGuard.php(51): Slim\Middleware\MethodOverride->call()
#20 $(SolutionDir)\userfrosting\middleware\usersession\UserSession.php(27): Slim\Extras\Middleware\CsrfGuard->call()
#21 $(SolutionDir)\userfrosting\vendor\slim\slim\Slim\Middleware\PrettyExceptions.php(67): UserFrosting\UserSession->call()
#22 $(SolutionDir)\userfrosting\vendor\slim\slim\Slim\Slim.php(1302): Slim\Middleware\PrettyExceptions->call()
#23 $(ProjectDir)\index.php(421): Slim\Slim->run()
#24 {main}
```

I'll try to find the bug and let you know how to fix it, unless you would be able to immediately tell how. It's likely one single line of code in the user creation function or a typo when assigning a variable or standard class value.

**Comment by alexweissman** [comment]
Hmm, this is odd.  The value for `secret_token` for a new user actually *should* be empty, and should only be generated "on demand" in the first event for which it is needed.  

I'm confused by this error because the column `secret_token` in table `user`, as a varchar, should have a default value of "".  However, I'm looking in the installation routine, and I see that this default is not explicitly set.

What database/version are you using?

**Comment by dorianmuthig** [comment]
I was running on MySQL 5.6.27 earlier today, but I upgraded everything to 5.7.9 around noon. The issue occurs with both.
Just to confirm, you are not setting the secret token on user creation like UserCake did? It's not important whether it's set or not, so long as a password reset or activation requests (except on initial registration) will set it to something else, if it is older than say, 10 minutes. But that's not really my concern here.
The table inspector in MySQL Workbench says the column `secret_token` is a non-nullable `varchar(32)` without a default value. It has implicit character encoding and collations set (table default) and features a comment explaining what it's for, but those things are not functionally relevant here. The non-nullable thing might be. If you want the value to not exist, making it nullable and adding a constraint that it must either be `NULL` or exactly 32 characters long and unique might be a good option. I'm not sure about support for such invalid data constraints in MySQL servers, and [the official documents](http://dev.mysql.com/doc/refman/5.7/en/constraint-invalid-data.html "MySQL :: MySQL 5.7 Reference Manual :: 1.8.3.3 Constraints on Invalid Data") aren't telling me much.
For now I just need it to work, so I can build a website with it. I can evaluate and fix the security issues later. I would just go ahead and set the default value on the column to an empty string, but I am concerned about what `UserFrosting\AccountController->activate()` or `UserFrosting\AccountController->setPassword()` and by inference `\Fortress\HTTPRequestFortress->validate()` do, if someone submits an empty validation or password reset token, and users with such tokens actually exist.

**Comment by alexweissman** [comment]
Yup, `secret_token` is changed every time a new account activity such as "reset password" or "activate account" is initiated.
It should be fine for it to default to empty.  The reason is that none of these activities will proceed unless the corresponding flag (`flag_verified`, `flag_password_reset`) is set to an appropriate value.  And, the flag is only set to the necessary value when the event is created (during which the `secret_token` is set as well).

**Comment by panchtox** [comment]
Hi, I'm experiencing the same issue when trying to register Master Account. I'm not an experienced programmer (much more an amateur one).
When I consult the error log I see:
`[Thu Nov 05 21:12:03.455335 2015] [:error] [pid 59416] [client 127.0.0.1:65386] Error in /Library/WebServer/Documents/#######/userfrosting/vendor/illuminate/database/Connection.php on line 624: SQLSTATE[HY000]: General error: 1364 Field 'secret_token' doesn't have a default value (SQL: insert into ``uf_user`` (``user_name``, ``display_name``, ``email``, ``password``, ``flag_verified``, ``locale``, ``primary_group_id``, ``title``, ``updated_at``, ``created_at``) values (#######, #######, ######@gmail.com, #######, 1, en_US, 1, New User, 2015-11-05 19:12:03, 2015-11-05 19:12:03)), referer: http://foo.example.com/#######/public/install/master
[Thu Nov 05 21:12:03.455462 2015] [:error] [pid 59416] [client 127.0.0.1:65386] #0`

Many thanks for your help.

**Comment by alexweissman** [comment]
Ok, for now just set that column in your database to a default value of "".

**Comment by alexweissman** [comment]
Alright, fixed in v0.3.1.5.
